Restrict search results based on user permissions [FC-0040]

pomegranited opened the pull request:
<!--

🌳🌳
🌳🌳🌳🌳         🌳 Note: Quince is in support. Fixes you make on master may still be needed on Quince.
    🌳🌳🌳🌳     If so, make another pull request against the open-release/quince.master branch,
🌳🌳🌳🌳         or ask in the #wg-build-test-release Slack channel if you have any questions or need help.
🌳🌳


Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply.
You may link to information rather than copy it, but only if the link is publicly
readable.  If you must linked information must be private (because it has secrets),
clearly label the link as private.

-->

## Description

Meilisearch allows us to create JWTs with embedded "filters", which guarantee that users searching with these tokens will only be permitted to see the documents that match these filters.

In Open edX, we grant permissions to block content based on its learning context, i.e. the course or content library that contains the block. Access to courses can be done in two ways, either individually on the course, or using `org` level rights. Access to content libraries is managed individually on the library (only library creation uses `org` level rights).

Because of these individual access rights, the list of courses and libraries a user has access to can be very long, >1000 in some instances (see [comment](https://github.com/openedx/modular-learning/issues/198#issuecomment-2027752922)). But since we need to encode that list into a JWT, we're limited to a total token size of 8Kib, but course keys can be very long (255 char). So to facilitate packing as many course/library identifiers into this token, this PR:

* Adds a SearchAccess model with a numeric ID, mapped to all CourseOverview and ContentLibrary context_keys.
* Adds an `access_id` field to the search-indexed Document, storing the SearchAccess.id relevant to that block's learning context.
* Adds an `access_id` filter to the JWT token generation, containing the SearchAccess.id's for the top 1,000 learning contexts that the user has access to. (We omit `access_id`s covered by the `org` filter described below.)
* Adds an `org`  filter to the JWT token generation, containing max 1,000 org names that the user has been granted access to.

## Supporting information

Content search is not enabled by default, so this change should have no impact on edX/2U or other sites running from master.

closes https://github.com/openedx/modular-learning/issues/198

Private-ref: [FAL-3692](https://tasks.opencraft.com/browse/FAL-3692)

## Testing instructions
Setup:

1. Ensure that you are running the latest https://github.com/open-craft/tutor-contrib-meilisearch -- this will ensure that `MEILISEARCH_ENABLED = True` is set in the CMS and the MFE.
1. Enable course-authoring MFE, running from the branch in https://github.com/openedx/frontend-app-course-authoring/pull/928
1. Load https://github.com/open-craft/taxonomy-sample-data/
1. Using this branch, run migrations: `python manage.py cms migrate`
1. Also using this branch, reindex studio content: `python manage.py cms reindex_studio --experimental`

Testing:

1. Login to Studio and the Course Authoring MFE as a global staff user.
1. Click the 'search' button in the header bar to start searching studio content.
    Ensure that you can see content from all your courses.
1. Repeat these steps with a non-global staff user who only has access to a single course or library.
    Ensure that user can only see that course/library when searching Studio content.

## Deadline

None

## Other information

1. Anyone using content.search will need to run a full reindex to add the new document `access_id` field:
   ```
   python manage.py cms reindex_studio --experimental
   ```

pomegranited committed:
feat: adds SearchAccess model

Stores a numeric ID for each course + library, which will generally be
shorter than the full context_key, so we can pack more of them into the
the Meilisearch search filter.

Also:

* Adds data migration pre-populates the SearchAccess model from the existing
  CourseOverview and ContentLibrary records
* Adds signal handlers to add/remove SearchAccess entries when content
  is created or deleted.
* Adds get_access_ids_for_request() helper method for use in views.
* Adds tests.

openedx-webhooks commented:
Thanks for the pull request, @pomegranited! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

pomegranited committed:
feat: use SearchAccess in documents and views

* Adds an access_id field to the document, which stores the
  SearchAccess.id for the block's context.
* Use the requesting user's allowed access_ids to filter search results
  to documents with those access_ids.
* Since some users have a lot of individual access granted, limit the
  number of access_ids in the filter to a large number (1_000)
* Updates tests to demonstrate.

pomegranited committed:
test: can't import content.search in lms tests

pomegranited committed:
Merge branch 'master' into jill/content-search-perms

pomegranited committed:
test: can't import content.search or content_staging in lms tests

rpenido committed:
fix: make access_id field filterable

rpenido commented on openedx/core/djangoapps/content/search/views.py:
```diff
@@ -29,6 +33,21 @@ def _get_meili_api_key_uid():
+    access_ids = get_access_ids_for_request(request)[:MAX_ACCESS_IDS_IN_FILTER]
+    return {
+        "filter": f"access_id IN {access_ids}",
+    }
```
Is it possible to separate "org level access" from "course level access", as stated in the issue? Like: `org IN ["orgA", "orgB", ...] OR access_id IN [1, 2, 3, ...]`

That way, we rely less on the `access_id` and its "limitations".

pomegranited replied:
Please see [3b07501](https://github.com/openedx/edx-platform/pull/34471/commits/3b075014b2aea8069c3900b407bd6622508a2a46)

rpenido reviewed (changes requested):
Hi, @pomegranited! LGTM :+1: 

I tested this, and it worked after a minor fix (e1e910a92058a8fa0f6961746113cf6893ce89a5). Could you make this change in your PR?

Also, I added a comment about filtering orgs.

Note that https://github.com/openedx/edx-platform/pull/34391 refactors part of the base code used here. If it gets merged before, you will have to move some things around here. I don't think it will be a lot, though.

pomegranited committed:
fix: use SearchAccess.get_or_create in signal handlers

In theory, we shouldn't have to do this, because the CREATE and DELETE
events should keep the SearchAccess table up-to-date.

But in practice, signals can be missed (or in tests, they may be
disabled). So we assume that it's ok to re-use a SearchAccess.id created
for a given course or library context_key.

pomegranited committed:
refactor: refactors the view tests to make them clearer

Uses helper methods and decorators to wrap the settings and patches used
by multiple view tests.

pomegranited commented on openedx/core/djangoapps/content/search/migrations/0001_initial.py:
```diff
@@ -0,0 +1,49 @@
+    for lib in ContentLibrary.objects.select_related("org").only("org", "slug"):
+        # ContentLibrary.library_key property is not available here, so we have to assemble it.
+        lib_key = LibraryLocatorV2(org=lib.org.short_name, slug=lib.slug)
+        SearchAccess.objects.create(context_key=lib_key)
```
If we're concerned about the performance of this data migration, I can remove it and add a management command instead.

pomegranited replied:
FYI @rpenido -- I decided to remove the data migration ([ef89591](https://github.com/openedx/edx-platform/pull/34471/commits/ef89591755a07d69aa8a4934240cfaeb180ff745)): running [`reindex_studio` handles this for us](https://github.com/open-craft/edx-platform/blob/jill/content-search-perms/openedx/core/djangoapps/content/search/documents.py#L83-L88).

pomegranited commented on openedx/core/djangoapps/content/search/handlers.py:
```diff
@@ -0,0 +1,35 @@
+
+
+@receiver(COURSE_CREATED)
+def create_course_search_access(course: CourseData, **kwargs):
```
Should these handlers be gated by `api.only_if_meilisearch_enabled` like in [edx-platform#34391](https://github.com/openedx/edx-platform/pull/34391/files#diff-0e43f240506a490b34aa601b067fd68ee73118a8158b5941133fe1f807cd47b6R31), to prevent this PR from affecting users not using this feature yet?

rpenido replied:
I think this is a good idea @pomegranited.

pomegranited replied:
Will do this after [edx-platform#34391](https://github.com/openedx/edx-platform/pull/34391/files#diff-0e43f240506a490b34aa601b067fd68ee73118a8158b5941133fe1f807cd47b6R31) merges.

bradenmacdonald replied:
Yes, but do we even need this handler? You have `get_or_create` in `_meili_access_id_from_context_key` so is there any advantage to pre-creating them here?

pomegranited replied:
Good point about the creation handlers, see [a28f3a9](https://github.com/openedx/edx-platform/pull/34471/commits/a28f3a93270cf7f79a8e04eb3924e7b567c195a6).

I left the deletion handlers in so that deleted courses/libraries don't take up any of the 1_000 slots.

bradenmacdonald replied:
> I left the deletion handlers in so that deleted courses/libraries don't take up any of the 1_000 slots.

Smart; I didn't think about that.

pomegranited committed:
feat: adds org filters to meilisearch filter

* Uses content_tagging.rules.get_user_orgs to fetch the user's
  content-related orgs for use in the meilisearch filter.
* Limits the number of orgs used to 1_000 to keep token size down

rpenido commented on openedx/core/djangoapps/content/search/tests/test_views.py:
```diff
@@ -1,14 +1,40 @@
+MOCK_API_KEY_UID = "3203d764-370f-4e99-a917-d47ab7f29739"
+
+
+def mock_meilisearch(enabled=True):
```
Nice approach here! :+1:

rpenido commented on openedx/core/djangoapps/content/search/views.py:
```diff
@@ -29,6 +35,26 @@ def _get_meili_api_key_uid():
+
+    # Everyone else is limited to their org roles...
+    user_orgs = [
+        org.short_name for org in get_user_orgs(request.user)[:MAX_ORGS_IN_FILTER]
```
If the user can access only one course in `orgA`, `orgA` will be returned here. I think we want to add only orgs that the user is an admin here. Maybe use `_get_content_creator_orgs` ?

pomegranited replied:
See [1181eea](https://github.com/openedx/edx-platform/pull/34471/commits/1181eeac843148f2b557eedb166a0143fb0c31a9)

rpenido commented on openedx/core/djangoapps/content/search/models.py:
```diff
@@ -0,0 +1,56 @@
+    """
+    Returns the SearchAccess.id values that the user has read access to.
+    """
+    courses, _ = get_courses_accessible_to_user(request)
```
I think we want to get the IDs only for course-level assignments here. If the user is staff in `orgA` and doesn't have any course-level role, this list should be empty. I'm not sure if we have a function for that.

Maybe use what we did in `_get_course_user_orgs` in the tagging rules, but return the course ids?

pomegranited replied:
Done with [6a3c30a](https://github.com/openedx/edx-platform/pull/34471/commits/6a3c30aa71b83e385e6f5e698187b33c7cbff1e3).

rpenido reviewed (commented):
Hi @pomegranited!

I think we went the wrong path here regarding the org + access_id filter.

If I understand correctly, we should use the access_id filter only for course-level role assignments.

I put some tests (that are falling) here (https://github.com/open-craft/edx-platform/pull/650) to show what I'm expecting. Could you check if my understanding is right? :smile:

PS: It adds a lot of boilerplate because I'm creating a course to test. You don't need to add these tests to this PR.

pomegranited committed:
Merge branch 'master' into jill/content-search-perms

pomegranited committed:
refactor: removes data migration

Users should use the reindex_studio management command to populate SearchAccess.

pomegranited committed:
refactor: adds functions to common.djangoapps.student.role_helpers

to allow general access to the user's RoleCache without having to access
private attributes of User or RoleCache.

Related changes:

* Moves some functionality from openedx.core.djangoapps.enrollments.data.get_user_roles
  to this new helper method.
* Use these new helper method in content_tagging.rules

pomegranited commented on openedx/core/djangoapps/content/search/models.py:
```diff
@@ -0,0 +1,56 @@
+from openedx.core.djangoapps.content_libraries.api import get_libraries_for_user
+
+
+class SearchAccess(models.Model):
```
1. Should this model be named something more general, e.g. `LearningContext`?
2. Should it live somewhere more generally accessible, e.g. one level up at `openedx.core.djangoapps.content`?

cf [relevant comments](https://github.com/openedx/modular-learning/issues/198#issuecomment-2043131883)

pomegranited replied:
@ormsbee @bradenmacdonald ^

ormsbee replied:
I'd just as soon as leave it here. I think there is a place for a more general model, but that's probably an ADR and some careful thinking since it's going to be used by _all the things_. So I'd rather have something more modestly scoped here for now.

pomegranited committed:
fix: get_access_ids_for_request only returns individual access

instead of all course keys that the user can read.

Org- and GlobalStaff access checks will handle the rest.

pomegranited committed:
fix: use org-level permissions when generating search filter

Also refactors tests to demonstrate this change for OrgStaff and
OrgInstructor users.

pomegranited commented:
> I put some tests (that are falling) here (https://github.com/open-craft/edx-platform/pull/650) to show what I'm expecting. Could you check if my understanding is right? 😄

Thank you for those added tests! I didn't use them directly, but please see 6a3c30aa71b83e385e6f5e698187b33c7cbff1e3 and c7215acf86d90a08ce0c62f95e5fe5d74c85e2aa for the relevant changes and tests.

bradenmacdonald commented on openedx/core/djangoapps/content/search/views.py:
```diff
@@ -29,6 +36,38 @@ def _get_meili_api_key_uid():
+    user_orgs = _get_user_orgs(request)[:MAX_ORGS_IN_FILTER]
+
+    # ...or the N most recent courses and libraries they can access.
+    access_ids = get_access_ids_for_request(request)[:MAX_ACCESS_IDS_IN_FILTER]
```
Do you think it's worth passing in the list of orgs so that we can exclude any courses for which the user has an org-level role? (Because we don't need to list the access_ids for any courses which are part of an org that's already included.) It would make this slightly more complex but I think it could be a worthwhile optimization.

pomegranited replied:
Great idea, thank you! It wasn't hard to do: [811687d](https://github.com/openedx/edx-platform/pull/34471/commits/811687d9f4271793bb786d9e2e31eca769b7c599)

rpenido reviewed (approved):
Thank you, @pomegranited! Good work here!
LGTM :+1: 

* [x] I tested this using the instructions in my tutor stack
* [x] I read through the code
* [ ] ~~I checked for accessibility issues~~
* [x] Includes documentation

pomegranited committed:
refactor: remove SearchAccess creation signal handlers

Lets SearchAccess entries be created on demand during search indexing.

pomegranited committed:
Merge branch 'master' into jill/content-search-perms

pomegranited committed:
feat: omit access_ids from the search filter that are covered by the user's org roles

bradenmacdonald reviewed (approved):
This is great, thanks. I will try to merge when I can but there are some unrelated issues affecting platform today so may need to wait.

openedx-webhooks commented:
@pomegranited 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/34471&entry.1484655318=https://github.com/pomegranited&entry.752974735=2024-04-04+07:17&entry.1917517419=2024-04-17+18:21&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
